src/main/scala/tech/cryptonomic/conseil/tezos/TezosDatabaseOperations.scala

```diff
@@ -156,64 +143,94 @@ object TezosDatabaseOperations {
-    * @param maybeFees  List of possible new fees that may have been calculated for different
-    *                   operation kinds.
-    * @return           Database rows
+    *
+    * @param fees      List of average fees for different operation kinds.
+    * @return          Database action possibly containing the number of rows written (if available from the underlying driver)
     */
-  def feesToDatabaseRows(maybeFees: List[Option[AverageFees]]): List[Tables.FeesRow] = {
-    maybeFees
-      .filter(_.isDefined)
-      .map{ someFee =>
-        val fee = someFee.get
-        Tables.FeesRow(
-          low = fee.low,
-          medium = fee.medium,
-          high = fee.high,
-          timestamp = fee.timestamp,
-          kind = fee.kind
-        )
-      }
-  }
-
+  def writeFeesIO(fees: List[AverageFees]): DBIO[Option[Int]] =
+    Tables.Fees ++= fees.map(RowConversion.ofAverageFees)
+  
   /**
     * Given the operation kind, return range of fees and timestamp for that operation.
     * @param kind  Operation kind
     * @return      The average fees for a given operation kind, if it exists
     */
-  def calculateAverageFees(kind: String): Option[AverageFees] = {
-    val operationKinds = Set[String]{kind}
-    val action = for {
-      o <- Tables.Operations
-      if o.kind.inSet(operationKinds)
-    } yield (o.fee, o.timestamp)
-    val op = dbHandle.run(action.distinct.sortBy(_._2.desc).take(numberOfFeesAveraged).result)
-    val results = Await.result(op, Duration.apply(awaitTimeInSeconds, SECONDS))
-    val resultNumbers = results.map(x => x._1.getOrElse("0").toDouble)
-    val m: Int = ceil(mean(resultNumbers)).toInt
-    val s: Int = ceil(stdev(resultNumbers)).toInt
-    results.length match {
-      case 0 => None
-      case _ =>
-        val timestamp = results.head._2
-        Some(AverageFees(max(m - s, 0), m, m + s, timestamp, kind))
+  def calculateAverageFeesIO(kind: String)(implicit ec: ExecutionContext): DBIO[Option[AverageFees]] = {
+    def computeAverage(ts: java.sql.Timestamp, fees: Seq[(Option[String], java.sql.Timestamp)]): AverageFees = {
+      val values = fees.map {
+        case (fee, _) => fee.map(_.toDouble).getOrElse(0.0)
+      }
+      val m: Int = ceil(mean(values)).toInt
+      val s: Int = ceil(stdev(values)).toInt
+      AverageFees(max(m - s, 0), m, m + s, ts, kind)
     }
-  }
 
+    val opQuery =
+      Tables.Operations
+        .filter(_.kind === kind)
+        .map(o => (o.fee, o.timestamp))
+        .distinct
+        .sortBy { case (_, ts) => ts.desc }
+        .take(numberOfFeesAveraged)
+        .result
+
+    opQuery.map { 
+      timestampedFees =>
+        timestampedFees.headOption.map {
+          case (_, latest) =>
+            computeAverage(latest, timestampedFees)
+        }
+    }
+  }
 
   /**
     * Delete all accounts in database not associated with block at maxLevel.
-    * @return No value, database update
+    * @return the number of rows removed
```

Comment:
Very nice, I don't know why I wrote otherwise

Reply:
it happens to me too, all the time